### ISSUE TITLE
GetIdentityServerPartialLoginAsync returns null

### ISSUE THREAD

**Issue by olee** [IdentityServer/IdentityServer3#2479]
Hi
I want to redirect my users to a registration page, when they try to authenticate with an external provider that is not linked to a local account yet. However I am unable to get the external identity in my registration controller.

I returned `new AuthenticateResult("/Account/Register", ctx.ExternalIdentity)` in my `UserService.AuthenticateExternalAsync` method like [here](https://github.com/IdentityServer/IdentityServer3.Samples/blob/master/source/CustomUserService/CustomUserService/ExternalRegistrationUserService.cs#L58), but when I try to get that partial login in my controller with `GetIdentityServerPartialLoginAsync()`, it returns null.

Any ideas on why this could happen?

Edit: I already confirmed that the `idsrv.partial` cookie was created

**Comment by brockallen** [comment]
You need to run in the same pipeline and after IdSvr so its cookie middleware can access the cookies and claims for you.

**Comment by olee** [comment]
I noticed the same thing.
The pipeline for IdentityServer is `/core`, however my controller uses `/Account` as to not collide with any IdentityServer features.
Is there a way to get it working with those two staying in their own sub-URLs?
When I tried that (using `[Route("/core")]` on the Account-controller), it caused errors.
I also tried setting `Path = "/"` in `CookieOptions`, however that broke stuff as well.
This whole thing is really confusing.

**Comment by brockallen** [comment]
We have some MVC samples that show running in the same pipeline. You do need to use attribute routing. https://github.com/IdentityServer/IdentityServer3.Samples/tree/master/source/CustomUserService

**Comment by olee** [comment]
Is there no way to have to cookies work on `/` instead of `/core`, so the controller can have a different subdomain?
Sorry for asking soo much - I'm not that familiar with that stuff yet.

**Comment by brockallen** [comment]
The cookie is created, protected, and validated by IdSvr. You need to run in its pipeline to have IdSvr provide you the user fromt he cookie.

**Comment by olee** [comment]
Would it work if I change the url `/Account/Register` to `/core/Account/Register` then?

**Comment by brockallen** [comment]
it should

**Comment by olee** [comment]
Thanks for your help!

**Comment by roblapp** [comment]
Is it possible to do a EULA without using MVC? I tried the above process and followed the example from the Sample repo. I did everything except use MVC and the GetIdentityServerPartialLoginAsync is still returning null. Instead of the EULA controller returning View() I simply return html content. The html content is a simple form which then gets submitted to my POST end point in the EULA controller via an ajax call. But I can't seem to get it working. Is this even possible? I dislike MVC and refuse to use it.

**Comment by brockallen** [comment]
> Is it possible to do a EULA without using MVC

Use whatever framework you want that runs in the pipeline.

**Comment by roblapp** [comment]
I should be more clear - I am using WebAPI controllers for my EULA controller. In my Startup.cs file I have the following code 

```
public void Configuration(IAppBuilder app)
{
            var config = new HttpConfiguration();
            config.MapHttpAttributeRoutes();
            app.UseWebApi(config);
            
            app.Map("/core", coreApp =>
            {
                       var identityServerOptions = new IdentityServerOptions
                       {
                             //... config IdentityServer exactly like CustomUserSample
                       }
                       coreApp.UseIdentityServer(identityServerOptions);
             }
}
```

Then in my custom UserServiceBase implementation I have this line that gets executed when the external user is not found in the local DB

`context.AuthenticateResult = new AuthenticateResult("~/eula", context.ExternalIdentity);`

My EULA controller looks like this

```
public class EndUserAgreementV1Controller : ApiController
{

	[HttpGet, Route("core/eula")]
	public async Task<IHttpActionResult> GetAsync()
	{
		var ctx = this.Request.GetOwinContext();
		var partialLogin = await ctx.Environment.GetIdentityServerPartialLoginAsync();
		if (partialLogin == null)
		{
			throw new Exception("partialLogin was null");
		}

		return new EndUserAgreementActionResult(); //Contains the HTML content which renders fine
	}

	[HttpPost, Route("core/eula")]
	public async Task<IHttpActionResult> PostAsync([FromBody] EndUserAgreementModel model)
	{
		var ctx = this.Request.GetOwinContext();
		var partialUser = await ctx.Environment.GetIdentityServerPartialLoginAsync();
		if (partialUser == null)
		{
			throw new ArgumentNullException(nameof(partialUser));
		}

		return new EndUserAgreementActionResult();
	}
}
```
The HTML page that is being returned with the GET request contains the following JavaScript
```
$(function() {
    var eulaEndPoint = "eula";

    $("#eula-yes-button")
        .click(function (e) {
            e.preventDefault();
            $.post(eulaEndPoint,
                {
                    isAccepted: true
                },
                function() {
                    console.log("after POST request");
                });
        });

    $("#eula-no-button")
        .click(function (e) {
            e.preventDefault();
            $.post(eulaEndPoint,
                {
                    isAccepted: false
                },
                function () {
                    console.log("after POST request");
                });
        });
});
```

The weird thing is that when the GET end point is called as a result of the CustomUserService re-direct, the ctx.Environment.GetIdentityServerPartialLoginAsync(); returns a NON null value. But when the user accepts the EULA and the POST end point is called then ctx.Environment.GetIdentityServerPartialLoginAsync();  returns null. What could possibly be happening?

**Comment by brockallen** [comment]
Your code must run in the same pipeline as IdentityServer, more specifically after IdentityServer.

**Comment by roblapp** [comment]
I have tried everything I know of. If I add

`app.UseWebApi(WebApiConfig.ConfigureWebApi());`

after

`app.Map("/core", coreApp =>{ ... }); `

I get a 404 when I try to call the core/eula end point. I also tried adding

```
app.Map("/core", webApi =>
            {
                webApi.UseWebApi(WebApiConfig.ConfigureWebApi());
            });
```

after the `app.Map("/core", coreApp =>{ ... }); ` initialization code and that had the same result (404's when calling core/eula). What am I missing? The only time I don't get 404's is when I add

`app.UseWebApi(WebApiConfig.ConfigureWebApi());`

before the `app.Map("/core", coreApp =>{ ... }); ` code, but then the GetPartialLoginAsync() method returns null (despite both being in <domain>/core)

**Comment by brockallen** [comment]
`Map` forks the pipeline, so anything outside and after the Map is a different branch. This is all just basic OWIN/Katana middleware at this point.

**Comment by roblapp** [comment]
Well, apparently I am lacking in OWIN/Katana knowledge. I suppose I will look into that a bit more. Thanks for the responses.

**Comment by brockallen** [comment]
My suggestion would be to prototype it without the IdentityServer middleware, then add that in.

**Comment by roblapp** [comment]
Brock - I will be honest... I don't know jack about OWIN/Katana - sorry for the ignorance. I did exactly what you said and that solved that problem! I just stripped things down and added them back one piece at a time. If anyone else has this issue this is what I did to get things working

In my Startup.cs file I have
```
app.Map("/core", coreApp =>
            {

                var identityServerOptions = new IdentityServerOptions
                {
                    //...
                };
                coreApp.UseIdentityServer(identityServerOptions);
                coreApp.UseWebApi(WebApiConfig.ConfigureWebApi());
            });
```

And then my EULA controllers have attribute routing marked with `[Route("eula")]`

In my custom UserServiceBase implementation I have this line being executed if the user is not yet in the local database

`context.AuthenticateResult = new AuthenticateResult("~/eula", context.ExternalIdentity);`

I will point some important things out...
1) The order in which WebApi is initialized in the app.Map lambda is VERY important. I found it MUST be initialized after UseIdentityServer has been set.

2) Do not mark controllers with [Route("core/eula")]. I overlooked this and that is why I was getting 404's.

A huge thanks to the legendary Brock Allen whose helpful feedback has saved me a few times!
